File: 18-Command/src/go/macro_command.go
package main

// MacroCommand groups commands (Command + Composite). Undo reverses them
// in REVERSE order so state unwinds correctly.
type MacroCommand struct {
	commands []Command
}

func NewMacroCommand(commands ...Command) *MacroCommand {
	return &MacroCommand{commands: commands}
}

func (m *MacroCommand) Execute() {
	for _, command := range m.commands {
		command.Execute()
	}
}

func (m *MacroCommand) Undo() {
	for i := len(m.commands) - 1; i >= 0; i-- {
		m.commands[i].Undo()
	}
}

File: 18-Command/src/rust/macro_command.rs
use std::rc::Rc;

use crate::command::Command;

/// Macro Command — groups commands (Command + Composite). undo reverses
/// them in REVERSE order so state unwinds correctly.
pub struct MacroCommand {
    commands: Vec<Rc<dyn Command>>,
}

impl MacroCommand {
    pub fn new(commands: Vec<Rc<dyn Command>>) -> Self {
        MacroCommand { commands }
    }
}

impl Command for MacroCommand {
    fn execute(&self) {
        for command in &self.commands {
            command.execute();
        }
    }

    fn undo(&self) {
        for command in self.commands.iter().rev() {
            command.undo();
        }
    }
}
